frontend/src/metabase-lib/queries/utils/dataset.ts

```diff
@@ -0,0 +1,70 @@
+import _ from "underscore";
+import type {
+  ConcreteFieldReference,
+  DatasetColumn,
+  DatasetData,
+  FieldReference,
+  TableColumnOrderSetting,
+} from "metabase-types/api";
+import { checkNotNull } from "metabase/lib/types";
+import Dimension from "metabase-lib/Dimension";
+
+export const datasetContainsNoResults = (data: DatasetData) =>
+  data.rows == null || data.rows.length === 0;
+
+/**
+ * Returns a MBQL field reference (FieldReference) for a given result dataset column
+ */
+export function fieldRefForColumn(column: DatasetColumn): FieldReference {
+  // NOTE: matching existing behavior of returning the unwrapped base dimension until we understand the implications of changing this
+  return (column.field_ref &&
+    checkNotNull(Dimension.parseMBQL(column.field_ref))
+      .baseDimension()
+      .mbql()) as FieldReference;
+}
+
+export function normalizeFieldRef(fieldRef: FieldReference) {
+  const dimension = Dimension.parseMBQL(fieldRef);
+  return dimension && dimension.mbql();
+}
+
+export function findColumnIndexForColumnSetting(
+  columns: DatasetColumn[],
+  columnSetting: TableColumnOrderSetting,
+) {
+  // NOTE: need to normalize field refs because they may be old style [fk->, 1, 2]
+  const fieldRef = normalizeFieldRef(checkNotNull(columnSetting.fieldRef));
+  // first try to find by fieldRef
+  if (fieldRef != null) {
+    const dimension = checkNotNull(Dimension.parseMBQL(fieldRef));
+    const index = _.findIndex(columns, col =>
+      dimension.isSameBaseDimension(
+        fieldRefForColumn(col) as ConcreteFieldReference,
+      ),
```

Comment:
Can we make `Dimension.prototype.isSameBaseDimension` accept `FieldReference` instead of `ConcreteFieldReference`? Then we wouldn't need the cast.

Reply:
[c53d002](https://github.com/metabase/metabase/pull/37700/commits/c53d002a76886efba8e61853a08633ddf9c9d9bb)

@kulyk confirmed we probably can use FieldReference, I don't see any harm in it